api/include/opentelemetry/common/stltypes.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <opentelemetry/version.h>
+
+#ifndef HAVE_STDLIB_CPP
+/* Private implementation of standard C++20 types */
+#include "opentelemetry/nostd/stl/span.h"
+#include "opentelemetry/nostd/stl/string_view.h"
+#include "opentelemetry/nostd/stl/variant.h"
+#include "opentelemetry/nostd/stl/utility.h"
+#else
+/* Standard library implementation */
+#include <cstddef>
+#include <variant>
+#include <memory>
+#include <string_view>
+#include <utility>
+
+#ifdef HAVE_GSL
+#include <gsl/gsl>
```

Comment:
I don't think we should be adding dependencies in the api, even optional ones. As above, we already have to maintain nostd/span, adding an ifdef here only adds complexity.

IMO there should be something outside of the API that makes it more convenient to map STL and GSL types into nostd ones when calling into otel.